File: 2023/1/2023-1.py
import re

tmp = [
    'one',
    'two',
    'three',
    'four',
    'five',
    'six',
    'seven',
    'eight',
    'nine'
]
tmp = dict(enumerate(tmp, 1))
DIGITS = dict(zip(tmp.values(), tmp.keys()))
DIGITS_RE = '|'.join(DIGITS.keys())  # 'one|two|three|...'


def insert_numeric_digit(text: str, first: bool) -> str:
    # first: True -- first digit, False -- last digit

    regex = '(' + DIGITS_RE + ')'
    if not first:
        # 'eat' everything before the last digit
        regex = '.*' + regex

    res = re.search(regex, text)
    if res:
        # insert the numeric digit at the digit word's beginning/end
        i = res.span()[0 if first else 1]
        text = text[:i] + str(DIGITS[res.group(1)]) + text[i:]

    return text


def outer_digits(text: str, use_text: bool = False) -> int:
    # add numeric digits for the first and the last text digits
    if use_text:
        text = insert_numeric_digit(text, True)
        text = insert_numeric_digit(text, False)

    d = re.findall(r'\d', text)
    return int(d[0] + d[-1])


with open('input.txt', 'r') as file:
    lines = file.readlines()

# use_t: False -- part 1, True -- part 2
for use_t in [False, True]:
    total = sum([outer_digits(line, use_t) for line in lines])
    print(f'Part {use_t + 1}: {total}')

File: 2023/1/2023-1.cpp
#include <iostream>
#include <fstream>
#include <vector>
// #include <algorithm>

const std::string NUMBERS{"0123456789"};
const std::vector<std::string> NUMBERS_CHAR{"one", "two", "three", "four",
    "five", "six", "seven", "eight", "nine"};

// Sums the first and last digits found in a string
int sum_outer(const std::string& line) {

    auto first = line.find_first_of(NUMBERS);
    if (first == std::string::npos) 
        return 0;

    std::string both(1, line[first]); 
    auto last = line.find_last_of(NUMBERS);
    both += line[last];
    return std::stoi(both);
}

int main() {

    std::ifstream ifile("input.txt"); 
    std::string line;
    long total1{0};
    long total2{0};

    while (ifile >> line) {
        total1 += sum_outer(line);

        // find first/last spelled digit
        size_t first_i, last_i;
        size_t first_pos{0}, last_pos{0};
        bool found{false};
        for (size_t i = 0; i < NUMBERS_CHAR.size(); ++i) {
            // first digit word
            auto pos = line.find(NUMBERS_CHAR[i]);
            if (pos != std::string::npos && (!found || pos < first_pos)) {
                found = true;
                first_i = i;
                first_pos = pos;
            }
            // last digit word
            pos = line.rfind(NUMBERS_CHAR[i]);
            if (pos != std::string::npos && pos > last_pos) {
                last_i = i;
                last_pos = pos;
            }
        }
        // insert digits
        if (found) {
            line.insert(first_pos, std::to_string(first_i + 1));
            //  + 1 because of the insertion at first_pos
            line.insert(last_pos + NUMBERS_CHAR[last_i].size() + 1, 
                std::to_string(last_i + 1));
        }

        total2 += sum_outer(line);
    }

    std::cout << "Part 1: " << total1 << "\n" <<
        "Part 2: " << total2 << "\n";
}